File: server-java-play-framework/app/controllers/ReservationApiController.java
package controllers;

import apimodels.AvailabilityResponse;
import java.math.BigDecimal;
import apimodels.ReservationResponse;
import apimodels.TimeSlotResponse;

import com.typesafe.config.Config;
import play.mvc.Controller;
import play.mvc.Result;
import play.mvc.Http;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;
import com.google.inject.Inject;
import java.io.File;
import play.libs.Files.TemporaryFile;
import openapitools.OpenAPIUtils;
import com.fasterxml.jackson.core.type.TypeReference;

import javax.validation.constraints.*;
import javax.validation.Valid;
import com.typesafe.config.Config;

import openapitools.OpenAPIUtils.ApiAction;

@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaPlayFrameworkCodegen", date = "2026-02-09T20:49:20.380387445Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
public class ReservationApiController extends Controller {
    private final ReservationApiControllerImpInterface imp;
    private final ObjectMapper mapper;
    private final Config configuration;

    @Inject
    private ReservationApiController(Config configuration, ReservationApiControllerImpInterface imp) {
        this.imp = imp;
        mapper = new ObjectMapper();
        this.configuration = configuration;
    }

    @ApiAction
    public Result createReservation(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuestartDate = request.getQueryString("startDate");
        Long startDate;
        if (valuestartDate != null) {
            startDate = Long.parseLong(valuestartDate);
        } else {
            startDate = null;
        }
        String valueendDate = request.getQueryString("endDate");
        Long endDate;
        if (valueendDate != null) {
            endDate = Long.parseLong(valueendDate);
        } else {
            endDate = null;
        }
        String valueofferId = request.getQueryString("offerId");
        Long offerId;
        if (valueofferId != null) {
            offerId = Long.parseLong(valueofferId);
        } else {
            offerId = null;
        }
        String valueofferLocationId = request.getQueryString("offerLocationId");
        Long offerLocationId;
        if (valueofferLocationId != null) {
            offerLocationId = Long.parseLong(valueofferLocationId);
        } else {
            offerLocationId = null;
        }
        String valueappKey = request.getQueryString("appKey");
        String appKey;
        if (valueappKey != null) {
            appKey = valueappKey;
        } else {
            appKey = null;
        }
        String valuemetaData = request.getQueryString("metaData");
        String metaData;
        if (valuemetaData != null) {
            metaData = valuemetaData;
        } else {
            metaData = null;
        }
        return imp.createReservationHttp(request, version, deviceId, accountId, startDate, endDate, offerId, offerLocationId, appKey, metaData);
    }

    @ApiAction
    public Result deleteReservation(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuereservationId = request.getQueryString("reservationId");
        Long reservationId;
        if (valuereservationId != null) {
            reservationId = Long.parseLong(valuereservationId);
        } else {
            throw new IllegalArgumentException("'reservationId' parameter is required");
        }
        return imp.deleteReservationHttp(request, version, reservationId, deviceId, accountId);
    }

    @ApiAction
    public Result reservableAvailability(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuereservableId = request.getQueryString("reservableId");
        Long reservableId;
        if (valuereservableId != null) {
            reservableId = Long.parseLong(valuereservableId);
        } else {
            throw new IllegalArgumentException("'reservableId' parameter is required");
        }
        String valuereservableType = request.getQueryString("reservableType");
        String reservableType;
        if (valuereservableType != null) {
            reservableType = valuereservableType;
        } else {
            throw new IllegalArgumentException("'reservableType' parameter is required");
        }
        String valueavailability = request.getQueryString("availability");
        String availability;
        if (valueavailability != null) {
            availability = valueavailability;
        } else {
            availability = null;
        }
        String valueavailabilitySummary = request.getQueryString("availabilitySummary");
        String availabilitySummary;
        if (valueavailabilitySummary != null) {
            availabilitySummary = valueavailabilitySummary;
        } else {
            availabilitySummary = null;
        }
        return imp.reservableAvailabilityHttp(request, version, reservableId, reservableType, deviceId, accountId, availability, availabilitySummary);
    }

    @ApiAction
    public Result searchAvailability(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuereservableId = request.getQueryString("reservableId");
        Long reservableId;
        if (valuereservableId != null) {
            reservableId = Long.parseLong(valuereservableId);
        } else {
            throw new IllegalArgumentException("'reservableId' parameter is required");
        }
        String valuereservableType = request.getQueryString("reservableType");
        String reservableType;
        if (valuereservableType != null) {
            reservableType = valuereservableType;
        } else {
            throw new IllegalArgumentException("'reservableType' parameter is required");
        }
        String valuestartDate = request.getQueryString("startDate");
        Long startDate;
        if (valuestartDate != null) {
            startDate = Long.parseLong(valuestartDate);
        } else {
            startDate = null;
        }
        String valueendDate = request.getQueryString("endDate");
        Long endDate;
        if (valueendDate != null) {
            endDate = Long.parseLong(valueendDate);
        } else {
            endDate = null;
        }
        String valuestart = request.getQueryString("start");
        Integer start;
        if (valuestart != null) {
            start = Integer.parseInt(valuestart);
        } else {
            start = 0;
        }
        String valuelimit = request.getQueryString("limit");
        Integer limit;
        if (valuelimit != null) {
            limit = Integer.parseInt(valuelimit);
        } else {
            limit = 100;
        }
        return imp.searchAvailabilityHttp(request, version, reservableId, reservableType, deviceId, accountId, startDate, endDate, start, limit);
    }

    @ApiAction
    public Result searchReservations(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueappKey = request.getQueryString("appKey");
        String appKey;
        if (valueappKey != null) {
            appKey = valueappKey;
        } else {
            appKey = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuefilterAccountId = request.getQueryString("filterAccountId");
        Long filterAccountId;
        if (valuefilterAccountId != null) {
            filterAccountId = Long.parseLong(valuefilterAccountId);
        } else {
            filterAccountId = null;
        }
        String valuereservableId = request.getQueryString("reservableId");
        Long reservableId;
        if (valuereservableId != null) {
            reservableId = Long.parseLong(valuereservableId);
        } else {
            reservableId = null;
        }
        String valuereservableType = request.getQueryString("reservableType");
        String reservableType;
        if (valuereservableType != null) {
            reservableType = valuereservableType;
        } else {
            reservableType = null;
        }
        String valuekeyword = request.getQueryString("keyword");
        String keyword;
        if (valuekeyword != null) {
            keyword = valuekeyword;
        } else {
            keyword = null;
        }
        String valuestartDate = request.getQueryString("startDate");
        Long startDate;
        if (valuestartDate != null) {
            startDate = Long.parseLong(valuestartDate);
        } else {
            startDate = null;
        }
        String valueendDate = request.getQueryString("endDate");
        Long endDate;
        if (valueendDate != null) {
            endDate = Long.parseLong(valueendDate);
        } else {
            endDate = null;
        }
        String valuestart = request.getQueryString("start");
        Integer start;
        if (valuestart != null) {
            start = Integer.parseInt(valuestart);
        } else {
            start = 0;
        }
        String valuelimit = request.getQueryString("limit");
        Integer limit;
        if (valuelimit != null) {
            limit = Integer.parseInt(valuelimit);
        } else {
            limit = 100;
        }
        return imp.searchReservationsHttp(request, version, deviceId, appKey, accountId, filterAccountId, reservableId, reservableType, keyword, startDate, endDate, start, limit);
    }

    @ApiAction
    public Result searchSchedule(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuereservableId = request.getQueryString("reservableId");
        Long reservableId;
        if (valuereservableId != null) {
            reservableId = Long.parseLong(valuereservableId);
        } else {
            throw new IllegalArgumentException("'reservableId' parameter is required");
        }
        String valuereservableType = request.getQueryString("reservableType");
        String reservableType;
        if (valuereservableType != null) {
            reservableType = valuereservableType;
        } else {
            throw new IllegalArgumentException("'reservableType' parameter is required");
        }
        String valuestartDate = request.getQueryString("startDate");
        Long startDate;
        if (valuestartDate != null) {
            startDate = Long.parseLong(valuestartDate);
        } else {
            throw new IllegalArgumentException("'startDate' parameter is required");
        }
        String valueendDate = request.getQueryString("endDate");
        Long endDate;
        if (valueendDate != null) {
            endDate = Long.parseLong(valueendDate);
        } else {
            throw new IllegalArgumentException("'endDate' parameter is required");
        }
        String valuetimeBucketMins = request.getQueryString("timeBucketMins");
        Integer timeBucketMins;
        if (valuetimeBucketMins != null) {
            timeBucketMins = Integer.parseInt(valuetimeBucketMins);
        } else {
            timeBucketMins = 30;
        }
        return imp.searchScheduleHttp(request, version, reservableId, reservableType, startDate, endDate, deviceId, accountId, timeBucketMins);
    }

}

File: server-scala-play-server/app/api/ReservationApiController.scala
package api

import org.openapitools.OpenApiExceptions
import javax.inject.{Inject, Singleton}
import play.api.libs.json._
import play.api.mvc._
import model.AvailabilityResponse
import model.ReservationResponse
import model.TimeSlotResponse

@javax.annotation.Generated(value = Array("org.openapitools.codegen.languages.ScalaPlayFrameworkServerCodegen"), date = "2026-02-11T19:31:43.310890579Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
@Singleton
class ReservationApiController @Inject()(cc: ControllerComponents, api: ReservationApi) extends AbstractController(cc) {
  /**
    * POST /api/3.18/reservation/create?deviceId=[value]&accountId=[value]&startDate=[value]&endDate=[value]&offerId=[value]&offerLocationId=[value]&appKey=[value]&metaData=[value]
    */
  def createReservation(): Action[AnyContent] = Action { request =>
    def executeApi(): Unit = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val startDate = request.getQueryString("startDate")
        .map(value => value.toLong)
        
      val endDate = request.getQueryString("endDate")
        .map(value => value.toLong)
        
      val offerId = request.getQueryString("offerId")
        .map(value => value.toLong)
        
      val offerLocationId = request.getQueryString("offerLocationId")
        .map(value => value.toLong)
        
      val appKey = request.getQueryString("appKey")
        
      val metaData = request.getQueryString("metaData")
        
      api.createReservation(deviceId, accountId, startDate, endDate, offerId, offerLocationId, appKey, metaData)
    }

    executeApi()
    Ok
  }

  /**
    * POST /api/3.18/reservation/delete?deviceId=[value]&accountId=[value]&reservationId=[value]
    */
  def deleteReservation(): Action[AnyContent] = Action { request =>
    def executeApi(): Unit = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val reservationId = request.getQueryString("reservationId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservationId", "query string")
        }
        
      api.deleteReservation(reservationId, deviceId, accountId)
    }

    executeApi()
    Ok
  }

  /**
    * POST /api/3.18/reservable/availability/update?deviceId=[value]&accountId=[value]&reservableId=[value]&reservableType=[value]&availability=[value]&availabilitySummary=[value]
    */
  def reservableAvailability(): Action[AnyContent] = Action { request =>
    def executeApi(): List[AvailabilityResponse] = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val reservableId = request.getQueryString("reservableId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservableId", "query string")
        }
        
      val reservableType = request.getQueryString("reservableType")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservableType", "query string")
        }
        
      val availability = request.getQueryString("availability")
        
      val availabilitySummary = request.getQueryString("availabilitySummary")
        
      api.reservableAvailability(reservableId, reservableType, deviceId, accountId, availability, availabilitySummary)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * GET /api/3.18/reservable/availability/search?deviceId=[value]&accountId=[value]&reservableId=[value]&reservableType=[value]&startDate=[value]&endDate=[value]&start=[value]&limit=[value]
    */
  def searchAvailability(): Action[AnyContent] = Action { request =>
    def executeApi(): List[AvailabilityResponse] = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val reservableId = request.getQueryString("reservableId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservableId", "query string")
        }
        
      val reservableType = request.getQueryString("reservableType")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservableType", "query string")
        }
        
      val startDate = request.getQueryString("startDate")
        .map(value => value.toLong)
        
      val endDate = request.getQueryString("endDate")
        .map(value => value.toLong)
        
      val start = request.getQueryString("start")
        .map(value => value.toInt)
        
      val limit = request.getQueryString("limit")
        .map(value => value.toInt)
        
      api.searchAvailability(reservableId, reservableType, deviceId, accountId, startDate, endDate, start, limit)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * GET /api/3.18/reservation/search?deviceId=[value]&appKey=[value]&accountId=[value]&filterAccountId=[value]&reservableId=[value]&reservableType=[value]&keyword=[value]&startDate=[value]&endDate=[value]&start=[value]&limit=[value]
    */
  def searchReservations(): Action[AnyContent] = Action { request =>
    def executeApi(): List[ReservationResponse] = {
      val deviceId = request.getQueryString("deviceId")
        
      val appKey = request.getQueryString("appKey")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val filterAccountId = request.getQueryString("filterAccountId")
        .map(value => value.toLong)
        
      val reservableId = request.getQueryString("reservableId")
        .map(value => value.toLong)
        
      val reservableType = request.getQueryString("reservableType")
        
      val keyword = request.getQueryString("keyword")
        
      val startDate = request.getQueryString("startDate")
        .map(value => value.toLong)
        
      val endDate = request.getQueryString("endDate")
        .map(value => value.toLong)
        
      val start = request.getQueryString("start")
        .map(value => value.toInt)
        
      val limit = request.getQueryString("limit")
        .map(value => value.toInt)
        
      api.searchReservations(deviceId, appKey, accountId, filterAccountId, reservableId, reservableType, keyword, startDate, endDate, start, limit)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * GET /api/3.18/reservable/schedule/search?deviceId=[value]&accountId=[value]&reservableId=[value]&reservableType=[value]&startDate=[value]&endDate=[value]&timeBucketMins=[value]
    */
  def searchSchedule(): Action[AnyContent] = Action { request =>
    def executeApi(): List[TimeSlotResponse] = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val reservableId = request.getQueryString("reservableId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservableId", "query string")
        }
        
      val reservableType = request.getQueryString("reservableType")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("reservableType", "query string")
        }
        
      val startDate = request.getQueryString("startDate")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("startDate", "query string")
        }
        
      val endDate = request.getQueryString("endDate")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("endDate", "query string")
        }
        
      val timeBucketMins = request.getQueryString("timeBucketMins")
        .map(value => value.toInt)
        
      api.searchSchedule(reservableId, reservableType, startDate, endDate, deviceId, accountId, timeBucketMins)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  private def splitCollectionParam(paramValues: String, collectionFormat: String): List[String] = {
    val splitBy =
      collectionFormat match {
        case "csv" => ",+"
        case "tsv" => "\t+"
        case "ssv" => " +"
        case "pipes" => "|+"
      }

    paramValues.split(splitBy).toList
  }
}
